bgpd/bgp_route.c

```diff
@@ -2983,8 +2990,8 @@ int bgp_update(struct peer *peer, struct prefix *p, uint32_t addpath_id,
 	bgp = peer->bgp;
 	rn = bgp_afi_node_get(bgp->rib[afi][safi], afi, safi, p, prd);
 	/* TODO: Check to see if we can get rid of "is_valid_label" */
-	if (afi == AFI_L2VPN && safi == SAFI_EVPN)
-		has_valid_label = (num_labels > 0) ? 1 : 0;
+	if (is_evpn_tunnel_type_vxlan(afi, safi, attr->encap_tunneltype))
+		has_valid_label = 1;
```

Comment:
Though not directly related to your change, I think the "has_valid_label" check has to be done by the callers, and then use that to pass down the "numlabels".

But related to your change, by just checking on encap being VxLAN, aren't you also including EVPN type-3 routes which don't have a label/VNI?

Reply:
Regarding has_valid_label, I like your idea. 

Yes. For type-3 routes, we would drop them when an "in" filter is applied. We felt that is in line with the expected behavior.

Reply:
Regarding evpn type 3 route, we decided to keep the behavior same for "in" filter as it is for "out" filter. If you feel type 3 routes should not be filtered out, I can make that change.

Reply:
If you can check and ensure we may never accidentally match for a type-3/type-4 or similar route which doesn't have a VNI/label against a "match evpn vni" statement, then we're ok.